source/core_mqtt_prop_deserializer.c

```diff
@@ -0,0 +1,1206 @@
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file core_mqtt_prop_deserializer.c
+ * @brief Implements the user-facing functions to de-serialize user-properties.
+ */
+#include <string.h>
+#include <inttypes.h>
+#include <stdbool.h>
+
+#include "core_mqtt_serializer.h"
+
+#include "private/core_mqtt_serializer_private.h"
+
+/* Include config defaults header to get default values of configs. */
+#include "core_mqtt_config_defaults.h"
+
+/**
+ * @brief Checks the parameters to #MQTT_PropertyReader_t and
+ * #MQTT_GetNextPropertyType.
+ *
+ * @param[in] mqttPropBuilder The property builder struct.
+ * @param[in] currentIndex The index of the property in the buffer.
+ *
+ * @return MQTTSuccess if all the checks pass;
+ */
+static inline MQTTStatus_t checkPropBuilderParams( MQTTPropBuilder_t * mqttPropBuilder,
+                                                   size_t * currentIndex );
+
+/*-----------------------------------------------------------*/
+
+static inline MQTTStatus_t checkPropBuilderParams( MQTTPropBuilder_t * mqttPropBuilder,
+                                                   size_t * currentIndex )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    if( ( mqttPropBuilder == NULL ) || ( mqttPropBuilder->pBuffer == NULL ) ||
+        ( currentIndex == NULL ) )
+    {
+        LogError( ( "Argument cannot be NULL." ) );
+        status = MQTTBadParameter;
+    }
+    else if( *currentIndex >= mqttPropBuilder->currentIndex )
+    {
+        LogWarn( ( "Index out of bounds." ) );
+        status = MQTTNoDataAvailable;
+    }
+    else
+    {
+        /* Nothing to do. All good. */
+    }
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+MQTTStatus_t MQTT_GetNextPropertyType( MQTTPropBuilder_t * pPropertyBuilder,
+                                       size_t * currentIndex,
+                                       uint8_t * property )
+{
+    MQTTStatus_t status = checkPropBuilderParams(pPropertyBuilder, currentIndex);
+
+    if( status != MQTTSuccess )
+    {
+        LogError( ( "Property type is invalid." ) );
+    }
+    else if( property == NULL )
+    {
+        LogError( ( "Argument cannot be NULL." ) );
+        status = MQTTBadParameter;
+    }
+    else
+    {
+        *property = pPropertyBuilder->pBuffer[ *currentIndex ];
```

Comment:
At this point the property output variable can be set to an invalid property and still be returned. Is that desirable?